### DISCUSSION TITLE
Unseal IDisposable implementations?

### DISCUSSION THREAD

**Discussion by baskren** [unoplatform/uno#13012] [General]
Recently, I've noticed that (inconsistently) the implementation of IDisposable methods (ex `void Dispose(bool disposing)`) is sealed on some platform implementations of some Framework Elements (today, I ran across this for the iOS implementation of `Border`). 

I've been using IDispose in ListView.ItemTemplate cell classes to terminate subscriptions that are used by the cell.  Yes, I probably could/should, with a bit of effort, use a better approach:

- using _weak events_
- using the **new** keyword (`new void Dispose(bool disposing)`) but I'm concerned that now (or in the future) this won't get called during view disposal
- refactor my view models:
  - refactor the collection class that contains each cell's view model to implement IDisposal 
  - extend the cell's view model with my additional cell decorator properties in question
  - subscribe to the event in question at the cell view model creation
  - unsubscribe to the event in question in the collection's `void Dispose(bool disposing)` method
  - use `Page.OnNavigatedFrom()` to dispose the view model, and thus the cell view model's subscriptions
- make a *really* complicated `Converter` and then `Bind` with that converter

The above are good solutions - but a lot more work than just being able to unsubscribe in the cell view's disposal.

Would it be possible to make the IDispose methods consistently  `virtual` or `protected`? 

Thank you for considering this.

**Comment by jeromelaban** [comment]
Thanks for the feedback. 

WinUI does not support disposing instances originally, which makes `IDisposable` a user-only contract that is never invoked by the framework. In general, using the finalizer and `IDisposable` is not a good idea for `UIElement`s, and it's generally best to work with the `Loaded`/`Unloaded` events, which are reliably invoked when `UIElement` instances are manipulated by the visual tree. If you explicitly know that your instances are meant to be disposed or cleaned up, you can use a similar pattern such as invoking `Cleanup()` on instances.

iOS/Android prevent the use of `IDisposable` because WinUI does not allow for it, but also because doing anything in the `Dispose` method can lead to a lot of memory, threading and stability issues for both platforms and it was decided a while back that allowing for the use of explicit memory management this way was more trouble than it was saving. As such, we will not restore the access to the `Dispose` method (at least not until the underlying framework provides it).

**Comment by baskren** [comment]
Thanks, @jeromelaban .  I appreciate the time you took to shed some light on this for me.  Very helpful explanation and confirmation of my suspicions.
